## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to import the movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [2]:
# Write your code for importing the csv file here
movies = pd.read_csv('C:\\Users\\russel\\Russel - Data Science\\5 - Libraries - Python\\Pandas\\Pandas DataFrames Datasets\\Movies.csv')
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:
# Write your code for inspection here
print(f"Shape of the dataframe: {movies.shape}")
print(f"Size of the dataframe: {movies.size}")

Shape of the dataframe: (3853, 28)
Size of the dataframe: 107884


In [4]:
movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [5]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

In [6]:
movies.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             3852.000000  3852.000000              3853.000000   
mean               163.046989   109.921080               784.507397   
std                123.927258    22.741728              3027.395958   
min                  1.000000    34.000000                 0.000000   
25%                 72.000000    95.000000                10.000000   
50%                134.000000   106.000000                58.000000   
75%                221.000000   120.000000               222.000000   
max                813.000000   330.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             3847.000000             3853.000000  3.853000e+03   
mean               747.290096             7575.558526  5.096547e+07   
std               1841.623945            15405.383114  6.932033e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                183.000000              721.000000  6.830957e+06   
50%                427.000000             1000.000000  2.790000e+07   
75%                685.000000            12000.000000  6.550000e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     3.853000e+03                3853.000000           3847.000000   
mean      1.024240e+05               11239.480145              1.378217   
std       1.502942e+05               18922.948757              2.056071   
min       5.000000e+00                   0.000000              0.000000   
25%       1.730900e+04                1817.000000              0.000000   
50%       5.052300e+04                3876.000000              1.000000   
75%       1.241850e+05               15972.000000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           3853.000000  3.853000e+03  3853.000000   
mean             326.720478  4.524817e+07  2003.061511   
std              407.920904  2.234208e+08    10.007168   
min                1.000000  2.180000e+02  1920.000000   
25%              102.000000  1.000000e+07  1999.000000   
50%              203.000000  2.400000e+07  2005.000000   
75%              391.000000  5.000000e+07  2010.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             3852.000000  3853.000000   3781.000000           3853.000000  
mean              1959.018432     6.462886      2.109569           9089.546068  
std               4472.171290     1.053843      0.353236          21277.375223  
min                  0.000000     1.600000      1.180000              0.000000  
25%                362.000000     5.900000      1.850000              0.000000  
50%                664.000000     6.600000      2.350000            206.000000  
75%                971.000000     7.200000      2.350000          11000.000000  
max             137000.000000     9.300000     16.000000         349000.000000

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [7]:
# Write you code here
print (movies.shape)

(3853, 28)


In [8]:
print (f"Number of rows: {movies.shape[0]}, Number of columns: {movies.shape[1]}")

Number of rows: 3853, Number of columns: 28


#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [9]:
# Write you code here

print(f"The number of columns having null values : {(movies.isnull().sum() > 0).sum()} columns.")

The number of columns having null values : 12 columns.


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  `color`
-  `director_facebook_likes`
-  `actor_1_facebook_likes`
-  `actor_2_facebook_likes`
-  `actor_3_facebook_likes`
-  `actor_2_name`
-  `cast_total_facebook_likes`
-  `actor_3_name`
-  `duration`
-  `facenumber_in_poster`
-  `content_rating`
-  `country`
-  `movie_imdb_link`
-  `aspect_ratio`
-  `plot_keywords`

In [10]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies_2 = movies.drop([
            "color", "director_facebook_likes", 
             "actor_1_facebook_likes", "actor_2_facebook_likes", 
             "actor_3_facebook_likes", "actor_2_name", "cast_total_facebook_likes", 
             "actor_3_name", "duration", "facenumber_in_poster", "content_rating", 
             "country", "movie_imdb_link", "aspect_ratio", "plot_keywords"
            ],axis=1)

In [11]:
movies_2.columns

Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes'],
      dtype='object')

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [12]:
print(f"The count of columns in the new dataframe : {len(movies_2.columns)}")

The count of columns in the new dataframe : 13


In [13]:
movies_2.shape[1]

13

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [14]:
# Write you code here
percentage_of_null = (movies_2.isnull().sum() / len(movies)) * 100
percentage_of_null

director_name             0.000000
num_critic_for_reviews    0.025954
gross                     0.000000
genres                    0.000000
actor_1_name              0.000000
movie_title               0.000000
num_voted_users           0.000000
num_user_for_reviews      0.000000
language                  0.103815
budget                    0.000000
title_year                0.000000
imdb_score                0.000000
movie_facebook_likes      0.000000
dtype: float64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [15]:
print(f"The column with the highest percentage of null values in New Dataframe : {percentage_of_null.idxmax()}")
print(f"The column with the highest percentage of null values in Old Dataframe : {round(100*(movies.isnull().sum()/len(movies.index)), 2).idxmax()}")

The column with the highest percentage of null values in New Dataframe : language
The column with the highest percentage of null values in Old Dataframe : aspect_ratio


**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [16]:
print(f"Before dropping duplicates, number of missing values : {movies['language'].isnull().sum()}")

Before dropping duplicates, number of missing values : 4


In [17]:
# Write your code for filling the NaN values in the 'language' column here
movies['language'] = movies['language'].fillna('English')

In [18]:
print(f"After dropping duplicates, number of missing values : {movies['language'].isnull().sum()}")

After dropping duplicates, number of missing values : 0


#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3675</font>
-  <font color='red'>3668</font>
-  <font color='red'>3674</font>

In [19]:
movies.loc[movies['language'].str.contains('english',case=False),'language'].count()

3675

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [20]:
movies_2.columns

Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes'],
      dtype='object')

In [21]:
# Before making changes

movies_2 [['gross','budget']].head(3)

gross       budget
0  760505847.0  237000000.0
1  309404152.0  300000000.0
2  200074175.0  245000000.0

In [22]:
# Write your code for unit conversion here
movies_2 [['gross','budget']] /= 1_000_000

In [23]:
# After making changes

movies_2 [['gross','budget']].head(3)

gross  budget
0  760.505847   237.0
1  309.404152   300.0
2  200.074175   245.0

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [24]:
# Write your code for creating the profit column here
movies_2 ['profit'] = movies_2 ['gross'] - movies_2 ['budget']

In [25]:
movies_2.head(2)

director_name  num_critic_for_reviews       gross  \
0   James Cameron                   723.0  760.505847   
1  Gore Verbinski                   302.0  309.404152   

                            genres actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi  CCH Pounder   
1         Action|Adventure|Fantasy  Johnny Depp   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   

   num_user_for_reviews language  budget  title_year  imdb_score  \
0                3054.0  English   237.0      2009.0         7.9   
1                1238.0  English   300.0      2007.0         7.1   

   movie_facebook_likes      profit  
0                 33000  523.505847  
1                     0    9.404152

In [26]:
# Write your code for sorting the dataframe here
movies_2.sort_values('profit',ascending=False,inplace=True)

In [27]:
movies_2.head()

director_name  num_critic_for_reviews       gross  \
0        James Cameron                   723.0  760.505847   
28     Colin Trevorrow                   644.0  652.177271   
25       James Cameron                   315.0  658.672302   
2704      George Lucas                   282.0  460.935665   
2748  Steven Spielberg                   215.0  434.949459   

                                genres         actor_1_name  \
0      Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28    Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                       Drama|Romance    Leonardo DiCaprio   
2704   Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                     Family|Sci-Fi         Henry Thomas   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
28                        Jurassic World            418214   
25                               Titanic            793059   
2704  Star Wars: Episode IV - A New Hope            911097   
2748          E.T. the Extra-Terrestrial            281842   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
28                  1290.0  English   150.0      2015.0         7.0   
25                  2528.0  English   200.0      1997.0         7.7   
2704                1470.0  English    11.0      1977.0         8.7   
2748                 515.0  English    10.5      1982.0         7.9   

      movie_facebook_likes      profit  
0                    33000  523.505847  
28                  150000  502.177271  
25                   26000  458.672302  
2704                 33000  449.935665  
2748                 34000  424.449459

In [28]:
top10 = movies_2.head(10)['movie_title']
top10

0                                          Avatar 
28                                 Jurassic World 
25                                        Titanic 
2704           Star Wars: Episode IV - A New Hope 
2748                   E.T. the Extra-Terrestrial 
16                                   The Avengers 
482                                 The Lion King 
230     Star Wars: Episode I - The Phantom Menace 
64                                The Dark Knight 
419                              The Hunger Games 
Name: movie_title, dtype: object

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

In [29]:
movies_2.head()['director_name']

0          James Cameron
28       Colin Trevorrow
25         James Cameron
2704        George Lucas
2748    Steven Spielberg
Name: director_name, dtype: object

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [30]:
top10.iloc[4]

'E.T. the Extra-Terrestrial\xa0'

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [31]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top = movies_2.loc [movies_2 ['num_voted_users'] > 25000].sort_values('imdb_score',ascending=False)

In [32]:
IMDb_Top_250 = IMDb_Top.iloc[:250]
IMDb_Top_250

director_name  num_critic_for_reviews       gross  \
1795        Frank Darabont                   199.0   28.341469   
3016  Francis Ford Coppola                   208.0  134.821952   
2543  Francis Ford Coppola                   149.0   57.300000   
64       Christopher Nolan                   645.0  533.316061   
325          Peter Jackson                   328.0  377.019252   
...                    ...                     ...         ...   
3674       Cristian Mungiu                   233.0    1.185783   
3767        John Carpenter                   318.0   47.000000   
2266        John Carpenter                   318.0   47.000000   
607           Michael Mann                   209.0   28.965197   
2708      David O. Russell                   410.0   93.571803   

                              genres      actor_1_name  \
1795                     Crime|Drama    Morgan Freeman   
3016                     Crime|Drama         Al Pacino   
2543                     Crime|Drama    Robert De Niro   
64       Action|Crime|Drama|Thriller    Christian Bale   
325   Action|Adventure|Drama|Fantasy     Orlando Bloom   
...                              ...               ...   
3674                           Drama  Anamaria Marinca   
3767                 Horror|Thriller  Jamie Lee Curtis   
2266                 Horror|Thriller  Jamie Lee Curtis   
607         Biography|Drama|Thriller         Al Pacino   
2708           Biography|Drama|Sport    Christian Bale   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
3674                   4 Months, 3 Weeks and 2 Days             44763   
3767                                      Halloween            157863   
2266                                      Halloween            157857   
607                                     The Insider            133526   
2708                                    The Fighter            275869   

      num_user_for_reviews  language  budget  title_year  imdb_score  \
1795                4144.0   English   25.00      1994.0         9.3   
3016                2238.0   English    6.00      1972.0         9.2   
2543                 650.0   English   13.00      1974.0         9.0   
64                  4667.0   English  185.00      2008.0         9.0   
325                 3189.0   English   94.00      2003.0         8.9   
...                    ...       ...     ...         ...         ...   
3674                 172.0  Romanian    0.59      2007.0         7.9   
3767                1191.0   English    0.30      1978.0         7.9   
2266                1191.0   English    0.30      1978.0         7.9   
607                  521.0   English   68.00      1999.0         7.9   
2708                 389.0   English   25.00      2010.0         7.9   

      movie_facebook_likes      profit  
1795                108000    3.341469  
3016                 43000  128.821952  
2543                 14000   44.300000  
64                   37000  348.316061  
325                  16000  283.019252  
...                    ...         ...  
3674                 14000    0.595783  
3767                 12000   46.700000  
2266                 12000   46.700000  
607                      0  -39.034803  
2708                 36000   68.571803  

[250 rows x 14 columns]

In [33]:
IMDb_Top_250.loc [:,'Rank'] = range(1,251)
IMDb_Top_250.head(3)

C:\Users\russel\AppData\Local\Temp\ipykernel_44840\484768559.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  IMDb_Top_250.loc [:,'Rank'] = range(1,251)


director_name  num_critic_for_reviews       gross       genres  \
1795        Frank Darabont                   199.0   28.341469  Crime|Drama   
3016  Francis Ford Coppola                   208.0  134.821952  Crime|Drama   
2543  Francis Ford Coppola                   149.0   57.300000  Crime|Drama   

        actor_1_name                movie_title  num_voted_users  \
1795  Morgan Freeman  The Shawshank Redemption           1689764   
3016       Al Pacino             The Godfather           1155770   
2543  Robert De Niro    The Godfather: Part II            790926   

      num_user_for_reviews language  budget  title_year  imdb_score  \
1795                4144.0  English    25.0      1994.0         9.3   
3016                2238.0  English     6.0      1972.0         9.2   
2543                 650.0  English    13.0      1974.0         9.0   

      movie_facebook_likes      profit  Rank  
1795                108000    3.341469     1  
3016                 43000  128.821952     2  
2543                 14000   44.300000     3

In [34]:
IMDb_Top_250.tail(3)

director_name  num_critic_for_reviews      gross  \
2266    John Carpenter                   318.0  47.000000   
607       Michael Mann                   209.0  28.965197   
2708  David O. Russell                   410.0  93.571803   

                        genres      actor_1_name   movie_title  \
2266           Horror|Thriller  Jamie Lee Curtis    Halloween    
607   Biography|Drama|Thriller         Al Pacino  The Insider    
2708     Biography|Drama|Sport    Christian Bale  The Fighter    

      num_voted_users  num_user_for_reviews language  budget  title_year  \
2266           157857                1191.0  English     0.3      1978.0   
607            133526                 521.0  English    68.0      1999.0   
2708           275869                 389.0  English    25.0      2010.0   

      imdb_score  movie_facebook_likes     profit  Rank  
2266         7.9                 12000  46.700000   248  
607          7.9                     0 -39.034803   249  
2708         7.9                 36000  68.571803   250

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [35]:
bucket_1 = IMDb_Top_250 [(IMDb_Top_250 ['imdb_score'] > 7.5) & (IMDb_Top_250 ['imdb_score'] < 8)].shape[0]
bucket_2 = IMDb_Top_250 [(IMDb_Top_250 ['imdb_score'] > 8) & (IMDb_Top_250 ['imdb_score'] < 8.5)].shape[0]
bucket_3 = IMDb_Top_250 [(IMDb_Top_250 ['imdb_score'] > 8.5) & (IMDb_Top_250 ['imdb_score'] < 9)].shape[0]
bucket_4 = IMDb_Top_250 [(IMDb_Top_250 ['imdb_score'] > 9) & (IMDb_Top_250 ['imdb_score'] < 9.5)].shape[0]
bucket_5 = IMDb_Top_250 [(IMDb_Top_250 ['imdb_score'] > 9.5) & (IMDb_Top_250 ['imdb_score'] < 10)].shape[0]

In [36]:
bucket_dictionary = {
        "Bucket_1" : bucket_1, 
        "Bucket_2" : bucket_2,
        "Bucket_3" : bucket_3,
        "Bucket_4" : bucket_4,
        "Bucket_5" : bucket_5,
}

In [37]:
print (f"The bucket with the maximum number of movies: {max(bucket_dictionary,key=bucket_dictionary.get)}")

The bucket with the maximum number of movies: Bucket_2


In [38]:
# Bucket_2 is 8 to 8.5

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [39]:
movies_2.columns

Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes', 'profit'],
      dtype='object')

In [40]:
# Write your code for creating three new dataframes here
# Include all movies in which Meryl_Streep is the lead

Meryl_Streep = movies_2 [movies_2['actor_1_name'].str.contains('Meryl Streep',case=False)] 

In [41]:
# Include all movies in which Leo_Caprio is the lead

Leo_Caprio = movies_2 [movies_2['actor_1_name'].str.contains('Leonardo DiCaprio',case=False)] 

In [42]:
# Include all movies in which Brad_Pitt is the lead

Brad_Pitt = movies_2 [movies_2['actor_1_name'].str.contains('Brad Pitt',case=False)] 

In [43]:
frames = [Meryl_Streep, Leo_Caprio, Brad_Pitt]

In [44]:
# Write your code for combining the three dataframes here

Combined = pd.concat(frames)
Combined

director_name  num_critic_for_reviews       gross  \
1322          David Frankel                   208.0  124.732962   
1471         Sydney Pollack                    66.0   87.100000   
1132            Nora Ephron                   252.0   94.125426   
1514          David Frankel                   234.0   63.536011   
392            Nancy Meyers                   187.0  112.703470   
2500         Phyllida Lloyd                   331.0   29.959436   
1784         Stephen Daldry                   174.0   41.597830   
2793          Robert Altman                   211.0   20.338609   
1038          Curtis Hanson                    42.0   46.815748   
1563          Carl Franklin                    64.0   23.209440   
1390         Robert Redford                   227.0   14.998070   
25            James Cameron                   315.0  658.672302   
94        Christopher Nolan                   642.0  292.568851   
859        Steven Spielberg                   194.0  164.435221   
283       Quentin Tarantino                   765.0  162.804648   
173   Alejandro G. Iñárritu                   556.0  183.635922   
431         Martin Scorsese                   490.0  127.968405   
347         Martin Scorsese                   352.0  132.373442   
49             Baz Luhrmann                   490.0  144.812796   
3025           Baz Luhrmann                   490.0  144.812796   
2484           Baz Luhrmann                   106.0   46.338728   
1336        Randall Wallace                    83.0   56.876365   
294         Martin Scorsese                   606.0  116.866727   
1364         Clint Eastwood                   392.0   37.304950   
246         Martin Scorsese                   267.0  102.608827   
1909             Jerry Zaks                    45.0   12.782508   
934             Danny Boyle                   118.0   39.778599   
1046             Sam Mendes                   323.0   22.877808   
1457              Sam Raimi                    63.0   18.636537   
312         Martin Scorsese                   233.0   77.679638   
608            Ridley Scott                   238.0   39.380442   
293            Edward Zwick                   166.0   57.366262   
383       Steven Soderbergh                   186.0  183.405771   
244              Doug Liman                   233.0  186.336103   
887             Neil Jordan                   120.0  105.264608   
448              David Ayer                   406.0   85.707116   
243       Steven Soderbergh                   198.0  125.531634   
2030  Alejandro G. Iñárritu                   285.0   34.300771   
2428         Andrew Dominik                   414.0   14.938570   
2602             Tony Scott                   122.0   12.281500   
2146    Angelina Jolie Pitt                   131.0    0.531009   
1397        Terrence Malick                   584.0   13.303319   
97            David Fincher                   362.0  127.490802   
646           David Fincher                   315.0   37.023395   
1608         Andrew Dominik                   273.0    3.904982   
579     Jean-Jacques Annaud                    76.0   37.901509   
749         Patrick Gilmore                    98.0   26.288320   
142       Wolfgang Petersen                   220.0  133.228348   
367              Tony Scott                   142.0    0.026871   

                                                 genres       actor_1_name  \
1322                               Comedy|Drama|Romance       Meryl Streep   
1471                            Biography|Drama|Romance       Meryl Streep   
1132                            Biography|Drama|Romance       Meryl Streep   
1514                               Comedy|Drama|Romance       Meryl Streep   
392                                Comedy|Drama|Romance       Meryl Streep   
2500                            Biography|Drama|History       Meryl Streep   
1784                                      Drama|Romance       Meryl Streep   
2793                                 Co

In [45]:
# Write your code for grouping the combined dataframe here
grouped_data = Combined.groupby('actor_1_name')[['num_critic_for_reviews','num_user_for_reviews']]

In [46]:
# Write the code for finding the mean of critic reviews and audience reviews here
grouped_data.mean()

num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [47]:
grouped_data.mean().sort_values('num_user_for_reviews',ascending=False)['num_user_for_reviews']

actor_1_name
Leonardo DiCaprio    914.476190
Brad Pitt            742.352941
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [48]:
grouped_data.mean().sort_values('num_critic_for_reviews',ascending=False)['num_critic_for_reviews']

actor_1_name
Leonardo DiCaprio    330.190476
Brad Pitt            245.000000
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64